src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenNullCheckedParameterTests.cs

```diff
@@ -0,0 +1,574 @@
+        }
+
+        [Fact]
+        public void TestNullCheckedSubstitution1()
+        {
+            var source = @"
+class A<T>
+{
+    internal virtual void M<U>(U u!) where U : T { }
+}
+class B1<T> : A<T> where T : class
+{
+    internal override void M<U>(U u!) { }
+}
+";
+            var compilation = CompileAndVerify(source);
+            compilation.VerifyIL("B1<T>.M<U>(U)", @"
+{
+      // Code size       15 (0xf)
+      .maxstack  1
+      IL_0000:  ldarg.1
+      IL_0001:  box        ""U""
+      IL_0006:  brtrue.s   IL_000e
+      IL_0008:  newobj     ""System.Exception..ctor()""
+      IL_000d:  throw
+      IL_000e:  ret
+}");
+        }
+
+        [Fact]
+        public void TestNullCheckedSubstitution2()
+        {
+            // PROTOTYPE : Should be error or warning here.
+            var source = @"
+class A<T>
+{
+    internal virtual void M<U>(U u!) where U : T { }
+}
+class B2<T> : A<T> where T : struct
+{
+    internal override void M<U>(U u!) { }
+}";
+            var compilation = CompileAndVerify(source);
+            compilation.VerifyIL("B2<T>.M<U>(U)", @"
+{
+    // Code size        1 (0x1)
+    .maxstack  0
+    IL_0000:  ret
+}");
+        }
+
+        [Fact]
+        public void TestNullCheckedSubstitution3()
+        {
+            var source = @"
+class A<T>
+{
+    internal virtual void M<U>(U u!) where U : T { }
+}
+class B3<T> : A<T?> where T : struct
+{
+    internal override void M<U>(U u!) { }
+}";
+            var compilation = CompileAndVerify(source);
+            compilation.VerifyIL("B3<T>.M<U>(U)", @"
+{
+    // Code size       15 (0xf)
+    .maxstack  1
+    IL_0000:  ldarg.1
+    IL_0001:  box        ""U""
+    IL_0006:  brtrue.s   IL_000e
+    IL_0008:  newobj     ""System.Exception..ctor()""
+    IL_000d:  throw
+    IL_000e:  ret
+}");
+        }
+
+        [Fact]
+        public void TestNullCheckedSubstitution4()
+        {
+            // PROTOTYPE : Should be error or warning here.
+            var source = @"
+class A<T>
+{
+    internal virtual void M<U>(U u!) where U : T { }
+}
+class B4 : A<int>
+{
+    internal override void M<U>(U u!) { }
+}";
+            var compilation = CompileAndVerify(source);
+            compilation.VerifyIL("B4.M<U>(U)", @"
+{
+    // Code size        1 (0x1)
+    .maxstack  0
+    IL_0000:  ret
+}");
+        }
+
+        [Fact]
+        public void TestNullCheckedSubstitution5()
+        {
+            var source = @"
+class A<T>
+{
+    internal virtual void M<U>(U u!) where U : T { }
+}
+class B5 : A<object>
+{
+    internal override void M<U>(U u!) { }
+}";
+            var compilation = CompileAndVerify(source);
+            compilation.VerifyIL("B5.M<U>(U)", @"
+{
+    // Code size       15 (0xf)
+    .maxstack  1
+    IL_0000:  ldarg.1
+    IL_0001:  box        ""U""
+    IL_0006:  brtrue.s   IL_000e
+    IL_0008:  newobj     ""System.Exception..ctor()""
+    IL_000d:  throw
+    IL_000e:  ret
+}");
+        }
```

Comment:
>} [](start = 8, length = 1)

It would be good to verify the execution of some of these cases, with `CompileAndVerify(source, expectedOutput: "...")`.

Perhaps test something like:
```C#
using System;
class Program
{
    static void M<T>(T t)
    {
        if (t is null) throw new ArgumentNullException();
    }
    static void Main()
    {
        Invoke(() => M(new object()));
        Invoke(() => M((object)null));
        Invoke(() => M((int?)1));
        Invoke(() => M((int?)null));
    }
    static void Invoke(Action a)
    {
        try
        {
            a();
            Console.WriteLine("ok");
        }
        catch (Exception e)
        {
            Console.WriteLine(e.GetType());
        }
    }
}
```